### PR TITLE
TrapsNeverHappen mode

### PR DESCRIPTION
**Pull request by kripken**
The goal of this mode is to remove obviously-unneeded code like
```wat
(drop
  (i32.load
    (local.get $x)))
```
In general we can't remove it, as the load might trap - we'd be removing
a side effect. This is fairly rare in general, but actually becomes quite
annoying with wasm GC code where such patterns are more common,
and we really need to remove them.

Historically the IgnoreImplicitTraps option was meant to help here. However,
in practice it did not quite work well enough for most production code, as
mentioned e.g. in #3934 . TrapsNeverHappen mode is an attempt to fix that,
based on feedback from @askeksa in that issue, and also I believe this
implements an idea that @fitzgen mentioned a while ago (sorry, I can't
remember where exactly...). So I'm hopeful this will be generally useful
and not just for GC.

The idea in TrapsNeverHappen mode is that traps are assumed to not
actually happen at runtime. That is, if there is a trap in the code, it will
not be reached, or if it is reached then it will not trap. For example, an
`(unreachable)` would be assumed to never be reached, which means
that the optimizer can remove it and any code that executes right before
it:

```wat
(if
  (..condition..)
  (block
    (..code that can be removed, if it does not branch out..)
    (..code that can be removed, if it does not branch out..)
    (..code that can be removed, if it does not branch out..)
    (unreachable)))
```

And something like a load from memory is assumed to not trap, etc.,
which in particular would let us remove that dropped load from earlier.

This mode should be usable in production builds with assertions
disabled, if traps are seen as failing assertions. That might not be true
of all release builds (maybe some use traps for other purposes), but
hopefully in some. That is, if traps are like assertions, then enabling
this new mode would be like disabling assertions in release builds
and living with the fact that if an assertion would have been hit then
that is "undefined behavior" and the optimizer might have removed
the trap or done something weird.

TrapsNeverHappen (TNH) is different from IgnoreImplicitTraps (IIT).
The old IIT mode would just ignore traps when computing effects.
That is a simple model, but a problem happens with a trap behind
a condition, like this:

```js
if (x != 0) foo(1 / x);
```

We won't trap on integer division by zero here only because of the
guarding if. In IIT, we'd compute no side effects on `1 / x`, and then
we might end up moving it around, depending on other code in
the area, and potentially out of the if - which would make it happen
unconditionally, which would break.

TNH avoids that problem because it does not simply ignore traps.
Instead, there is a new `hasUnremovableSideEffects()` method
that must be opted-in by passes. That checks if there are no side
effects, or if there are, if we can remove them - and we know we can
remove a trap if we are running under TrapsNeverHappen mode,
as the trap won't happen by assumption. A pass must only use that
method where it is safe, that is, where it would either remove the
side effect (in which case, no problem), or if not, that it at least does
not move it around (avoiding the above problem with IIT).

This PR does not implement all optimizations possible with
TNH, just a small initial set of things to get started. It is already
useful on wasm GC code, including being as good as IIT on removing
unnecessary casts in some cases, see the test suite updates here.
Also, a significant part of the 18% speedup measured in

https://github.com/WebAssembly/binaryen/pull/4052#issuecomment-893078019

is due to my testing with this enabled, as otherwise the devirtualization
there leaves a lot of unneeded code.

### CODE REVIEW

**Comment by kripken** [comment]
Summarizing some offline discussion: One concern is that builds may actually need a way to have runtime errors even in release builds with assertions disabled. And this mode will try to remove every single trap from the code as best it can. It seems like if code wants such a runtime error, it could:

1. Throw a wasm exception.
2. Call an import to error (on the web, one line of JS can throw).
3. Not use this option.

If those aren't good enough, then it seems like we'd need in Binaryen IR not just an Unreachable instruction but also one like it that is meant to not be removed in this mode. That seems like more than we need here, I suspect.

Or, we could treat a plain Unreachable different than a possible trap inside another instruction - as the main goal here is to remove `(drop (load))`, etc.,  and not `(unreachable)`. But that is tricky, as we may lower instructions into things that contain a plain Unreachable (where that would have the same observable behavior but is smaller, for example). So that would add complexity, and is another issue with the existing IIT mode - Ignore Implicit Traps does not ignore the "non-implicit" plain Unreachable, that would be good to fix with this new mode.

Overall, we likely don't need this for LLVM-using languages, as LLVM already removes code using UB. For new GC languages it seems useful, and I hope they can design to use one of options 1-2 in the list above, if they do need some runtime traps to remain.

**Comment by MaxGraey** [comment]
Can I ask you a question? This new mode is able to remove dead loads (drop (load ...)) and allows for example to convert 0 / x to 0 without considering that x can be in theory zero? But it does not remove explicit `unreachable` ops? So in essence it behaves like LLVM / GCC with such cases?

**Comment by kripken** [comment]
> But it does not remove explicit unreachable ops? 

Actually no, as proposed it *does* potentially remove explicit unreachable instructions. This PR doesn't yet, but the design is "traps can be removed", and an `(unreachable)` traps, so it can be.

This is like LLVM's unreachable, which indicates undefined behavior, I believe. And different from LLVM's trap, which is guaranteed to trap and cannot be removed.

**Comment by MaxGraey** [comment]
Hmm, I see. Thanks for the explanation

**Comment by aheejin** [comment]
Sorry for being late to the discussion, and a basic question: 

Why can't we remove the `load` in your snippet with `iit` mode?
```wasm
(drop
  (i32.load
    (local.get $x)))
```
The `load` here is assumed to not trap, then can't we remove the whole thing? I haven't read all of #3934 though.

I understand that `iit` has an important downside though: that it is generally not safe to unconditionalize a trapping instruction guarded by a condition but `iit` does that, and this new mode wouldn't do that. But the code snippet above with `drop` and `load` doesn't have any condition and I'm wondering why we can't remove it with `iit`.

**Comment by kripken** [comment]
Sorry if that wasn't clear - a load like that can be removed in both IIT and TNH modes. (The difference between them only happens on other things, like the safety issue you mention, and that TNH can remove code before a definite trap.)

**Comment by dschuff** [comment]
@MaxGraey I think it would be really useful to get some feedback about this from non-C users of Binaryen about this over time, to see whether we can get good optimization opportunities out of practical code, as well as whether this kind of behavior is easy to generate code for, or whether it becomes problematic.

One other thing we discussed offline is that it probably would make sense to go ahead and replace `iit` mode with some flavor of this, (since it's off by default and doing unsound optimizations) but keep it as an experimental feature for a while and see what the costs and benefits could be.

**Comment by MaxGraey** [comment]
@dschuff I absolutely agree that we need some mode to get rid of "dead loads" and other dead expressions that potentially contain implicit traps, but since the expression either never gets reached or the result of such expression is never used (case with dropped load), in my opinion it would be quite safe to turn on such mode even by default. But I'm not so sure about UBs removal, 0 / x -> 0 conversions and other things rather common in LLVM. Maybe it makes sense to have it as a separate mode turned off by default. Binaryen already have something similar for floating arithmetic (fast-float mode)

**Comment by kripken** [comment]
In offline discussion the general plan is to land this while mentioning it is still experimental, then suggest people using the existing IIT use it, then eventually remove IIT, and meanwhile to experiment and improve this mode.

File: src/ir/effects.h [hunk 1]
```diff
@@ -143,10 +145,37 @@ class EffectAnalyzer {
...
+  //  1. Remove any side effects present, if any, so they no longer exists.
+  //  2. Keep the code exactly where it is.
+  //
+  // If instead of 1&2 a pass kept the side effect and also reordered the code
+  // with other things, then that could be bad, as the side effect might have
+  // been behind a condition that avoids it occurring.
+  //
+  // TODO: Go through the optimizer and use this in all places that do not move
+  //       code around.
+  bool hasUnremovableSideEffects() const {
```
Comment (tlively) [inline src/ir/effects.h:175]: I assume the common case of usage will be `if (!hasUnremovableSideEffects) { /* remove side effects */ }`. Would it be worth eliminating that double negative by changing this to `canRemoveSideEffects` or even `canRemove`?

File: test/lit/passes/optimize-instructions-gc-iit.wast [hunk 2]
```diff
@@ -3,16 +3,22 @@
 ;; RUN:   | filecheck %s
 ;; RUN: wasm-opt %s --optimize-instructions --ignore-implicit-traps --enable-reference-types --enable-gc --nominal -S -o - \
 ;; RUN:   | filecheck %s --check-prefix NOMNL
+;; Also test trapsNeverHappen (with nominal; no need for both type system modes).
+;; RUN: wasm-opt %s --optimize-instructions --traps-never-happen --enable-reference-types --enable-gc --nominal -S -o - \
+;; RUN:   | filecheck %s --check-prefix TNHNL
```
Comment (tlively) [inline test/lit/passes/optimize-instructions-gc-iit.wast:8]: I would probably go with NOMNL-TNH or something even more verbose here to prioritize clarity over matching character counts.

**Review by tlively — COMMENTED** [review]
I'm very excited about this change, especially for future CFG-aware optimizations.

File: src/ir/effects.h (hunk 1 above)
Comment (kripken) [reply src/ir/effects.h:175]: Not necessarily, I think. One of the two usages already present is the opposite,
```
// Check for various things that make us fail.
if (some check) return;
if (hasUnremovableSideEffects()) return;
if (other check) return;
// All good, do the opt here.
```
Basically we replace `hasSideEffects()` with `hasUnremovableSideEffects()`. I think keeping those two methods similar is easier to keep in one's head, personally.

File: test/lit/passes/optimize-instructions-gc-iit.wast (hunk 2 above)
Comment (kripken) [reply test/lit/passes/optimize-instructions-gc-iit.wast:8]: Sounds good, done.

**Review by tlively — APPROVED** [review]

**Pull request merged**